sqlmesh/core/engine_adapter/base.py

```diff
@@ -748,6 +749,153 @@ def _merge(
+            .with_(
+                "static",
+                exp.Select()
+                .from_(target_table)
+                .select(*columns_to_types)
+                .where(f"{valid_to_name} IS NOT NULL"),
+            )
+            # Latest Records that can be updated
+            .with_(
+                "latest",
+                exp.select(*columns_to_types).from_(target_table).where(f"{valid_to_name} IS NULL"),
+            )
+            # Deleted records which can be used to determine `valid_from` for undeleted source records
+            .with_(
+                "deleted",
+                exp.select(*[f"static.{col}" for col in columns_to_types])
+                .from_("static")
+                .join("latest", using=unique_key, join_type="left")
+                .where(f"latest.{valid_to_name} IS NULL"),
+            )
+            # Get the latest `valid_to` deleted record for each unique key
+            .with_(
+                "latest_deleted",
+                exp.select(
+                    *unique_key,
+                    f"MAX({valid_to_name}) AS {valid_to_name}",
+                )
+                .from_("deleted")
+                .group_by(*unique_key),
+            )
+            # Do a full join between latest records and source table in order to combine them together
+            .with_(
+                "joined",
+                exp.select(
+                    *(f"latest.{col} AS t_{col}" for col in columns_to_types),
+                    *(f"source.{col} as s_{col}" for col in unmanaged_columns),
+                )
+                .from_("latest")
+                .join("source", using=unique_key, join_type="full"),
+            )
+            # Get deleted, new, no longer current, or unchanged records
+            .with_(
+                "non_updated_rows",
+                exp.select(
+                    *(f"COALESCE(t_{col}, s_{col}) as {col}" for col in unmanaged_columns),
+                    f"""
+                    CASE 
+                        WHEN t_{valid_from_name} IS NULL 
+                             AND latest_deleted.{unique_key[0]} IS NOT NULL 
+                        THEN CASE 
+                                WHEN latest_deleted.{valid_to_name} > s_{updated_at_name} 
+                                THEN latest_deleted.{valid_to_name} 
+                                ELSE s_{updated_at_name} 
+                             END 
+                        WHEN t_{valid_from_name} IS NULL 
+                        THEN {start_time} 
+                        ELSE t_{valid_from_name} 
+                    END AS {valid_from_name}""",
+                    f"""
+                    CASE 
+                        WHEN s_{updated_at_name} > t_{updated_at_name} 
+                        THEN s_{updated_at_name} 
+                        WHEN s_{unique_key[0]} IS NULL 
+                        THEN CAST('{to_ts(execution_time)}' AS TIMESTAMP) 
```

Comment:
why ts and the previous with timestamp? just do 1970-01-01 and ds?

Reply:
I find it easier to follow since we are wanting a timestamp to provide a timestamp. Yes it is correct that we could provide a date and cast it to timestamp I don't see the advantage to that.